source/Irrlicht/CBAWMeshWriter.cpp

```diff
@@ -0,0 +1,269 @@
+			nearestFree += sizeof(core::aabbox3df);
+			*((uint64_t *)nearestFree) = reinterpret_cast<uint64_t>(_obj->getMeshDataAndFormat());
+			nearestFree += sizeof(uint64_t);
+			*((video::E_INDEX_TYPE *)nearestFree) = _obj->getIndexType();
+			nearestFree += sizeof(video::E_INDEX_TYPE);
+			*((int32_t *)nearestFree) = _obj->getBaseVertex();
+			nearestFree += sizeof(int32_t);
+			*((uint64_t *)nearestFree) = _obj->getIndexCount();
+			nearestFree += sizeof(uint64_t);
+			*((size_t *)nearestFree) = _obj->getIndexBufferOffset();
+			nearestFree += sizeof(size_t);
+			*((size_t *)nearestFree) = _obj->getInstanceCount();
+			nearestFree += sizeof(size_t);
+			*((uint32_t *)nearestFree) = _obj->getBaseInstance();
+			nearestFree += sizeof(uint32_t);
+			*((E_PRIMITIVE_TYPE *)nearestFree) = _obj->getPrimitiveType();
+			nearestFree += sizeof(E_PRIMITIVE_TYPE);
+			*((E_VERTEX_ATTRIBUTE_ID *)nearestFree) = _obj->getPositionAttributeIx();
+			nearestFree += sizeof(E_VERTEX_ATTRIBUTE_ID);
+
+			core::XXHash_256(data, m_headers[_headerIdx].blobSizeDecompr, m_headers[_headerIdx].blobHash);
+			_file->write(data, m_headers[_headerIdx].blobSizeDecompr);
+			calcAndPushNextOffset(!_headerIdx ? 0 : m_headers[_headerIdx-1].blobSizeDecompr);
+
+			std::free(data);
+		}
+		template<>
+		void CBAWMeshWriter::exportAsBlob<video::IVirtualTexture>(video::IVirtualTexture* _obj, uint32_t _headerIdx, io::IWriteFile* _file)
+		{
+			video::ITexture* tex;
+			if (!(tex = dynamic_cast<video::ITexture*>(_obj)))
+				return;
+			//! @todo @bug here we get absolute path and we want relative
+			const io::path path = tex->getName().getInternalName();
+			const uint32_t len = std::strlen(path.c_str()) + 1;
+
+			m_headers[_headerIdx].blobSize = m_headers[_headerIdx].blobSizeDecompr = len;
+			core::XXHash_256(path.c_str(), len, m_headers[_headerIdx].blobHash);
+			calcAndPushNextOffset(!_headerIdx ? 0 : m_headers[_headerIdx-1].blobSizeDecompr);
+
+			_file->write(path.c_str(), len);
+		}
+		template<>
+		void CBAWMeshWriter::exportAsBlob<IMeshDataFormatDesc<core::ICPUBuffer> >(IMeshDataFormatDesc<core::ICPUBuffer>* _obj, uint32_t _headerIdx, io::IWriteFile* _file)
+		{
+			uint8_t* const data = (uint8_t*)std::malloc(m_headers[_headerIdx].blobSizeDecompr);
+
+			E_COMPONENTS_PER_ATTRIBUTE* cpa = (E_COMPONENTS_PER_ATTRIBUTE*)data;
+			E_COMPONENT_TYPE* attType = (E_COMPONENT_TYPE*)((uint8_t*)cpa + EVAI_COUNT * sizeof(E_COMPONENTS_PER_ATTRIBUTE));
+			size_t* attStride = (size_t*)((uint8_t*)attType + EVAI_COUNT * sizeof(E_COMPONENT_TYPE));
+			size_t* attOffset = (size_t*)((uint8_t*)attStride + EVAI_COUNT * sizeof(size_t));
+			uint32_t* attDiv = (uint32_t*)((uint8_t*)attOffset + EVAI_COUNT * sizeof(size_t));
+			uint64_t* attBufs = (uint64_t*)((uint8_t*)attDiv + EVAI_COUNT * sizeof(uint32_t));
+
+			uint64_t* idxBuf = (uint64_t*)((uint8_t*)attBufs + EVAI_COUNT * sizeof(uint64_t));
+
+			for (int i = 0; i < EVAI_COUNT; ++i)
```

Comment:
Loop unrolling will help and happen automagically by the compiler... but!

The memory is being walked very erratically, which the compiler won't fix, its much better for performance to have many loops for each array to be saved.